osu.Game/Screens/Select/Carousel/CarouselBeatmapSet.cs

```diff
@@ -126,6 +126,8 @@ private int compareUsingAggregateMax(CarouselBeatmapSet other, Func<BeatmapInfo,
             return ValidBeatmaps.Max(func).CompareTo(other.ValidBeatmaps.Max(func));
         }
 
+        private int fallbackWhenZero(CarouselBeatmapSet other, int i) => (i == 0) ? -other.BeatmapSet.DateAdded.CompareTo(BeatmapSet.DateAdded) : i;
```

Comment:
There's already a fallback for `CarouselItem`s that is supposed to handle this.
https://github.com/ppy/osu/blob/fc7fc3d673cf0a843d4eeea0225bcde76b987c36/osu.Game/Screens/Select/Carousel/CarouselGroup.cs#L89-L96

Have you checked why it doesn't work?

Reply:
_I tested this using VS Code Debug._

The test consists of two beatmaps:
1. First
2. Second

When I started the game, the two beatmaps had an `ItemID` of 0. The beatmaps are then assigned `ItemID`'s in `CarouselGroup.AddItem`:
1. First is assigned with `ItemID` of 1
2. Second is assigned with `ItemID` of 2

After playing with First, First's `ItemID` is reassigned to 3. Now:
1. First is assigned with `ItemID` of 3
2. Second is assigned with `ItemID` of 2

Which may be the reason First gets placed after Second.

Reply:
That sounds completely broken, and should be the focus of fixing this issue.

Reply:
@frenzibyte well spotted. Directly offending call stack:

![image](https://user-images.githubusercontent.com/20418176/210132520-f4066ed7-da64-47ce-88b5-e6dd3f76360a.png)

That call stack is obscured as the `UpdateBeatmapSet()` call is scheduled. Call site at point of scheduling:

![image](https://user-images.githubusercontent.com/20418176/210132579-2687b5e2-58b8-4557-9273-61c5023b9bf9.png)

i.e. a realm subscription callback is triggering this. This curiously happens at the *player loader*. The direct culprit is 99% this:

https://github.com/ppy/osu/blob/ceedc3f93469912efd3b66eb5c30a441bd8a9f56/osu.Game/Screens/Play/SubmittingPlayer.cs#L146-L151

Dunno where to take this next, really. Everything is sorta behaving to expectation but the end result is that last play date is inadvertently used as the tiebreaker.

Reply:
Since the `ItemID`'s starts from 1 to the number of `items`. How about just finding a missing `ItemID` and then assigning it?

I tried this code and it works.
``` C#
var itemsHaveMissingItemID = true;
ulong missingItemID = 1;

foreach (var item in items)
{
    if (item.ItemID != missingItemID)
    {
        i.ItemID = missingItemID;
        itemsHaveMissingItemID = false;
    }

   missingItemID++;
}

if (!itemsHaveMissingItemID)
i.ItemID = ++currentItemID;
```

It replaces:
https://github.com/ppy/osu/blob/ceedc3f93469912efd3b66eb5c30a441bd8a9f56/osu.Game/Screens/Select/Carousel/CarouselGroup.cs#L42

Reply:
Doesn't seem like a particularly good solution. There's really no guarantee that the gap is caused by updating the same beatmap.

Reply:
How about using `DateAdded` instead of `ItemID`?

Reply:
That is what I'd lean towards, if possible.